### PR TITLE
Implement MQTT subscription and event handling

### LINKED ISSUE #38: Implement MQTT subscription
## Description

The MQTT controller needs to be expanded to support subscriptions and the handling of those events.
i.e sending the relevant information from the subscribed topics to wherever they need to be in the app.

## Checklist

- [x] MQTT support subscription
- [x] Android studio can now subscribe to certain topics
- [x] The intro screen subscribe button is used to subscribe to topics

[PArt of Epic Feature 5](https://github.com/DIT113-V22/group-04/wiki/EF-5-Connectivity)

### PR DESCRIPTION
**Pull request by Soarnir**
# Description

This solution covers subscribing textview objects to topics, where they are then handled by the MQTT controller.
Due to constraints regarding activity classes and the lack of permanence present when changing scenes, the subscription map must be cleared upon class initialisation to prevent bloat.
The text objects subscribe to a topic by using the `update(TextView text, String topic)` method present in the MQTT controller.
They are then provided relevant messages by the callback function which is initialised upon controller connection.

## Issues

- Closes #38 

## Dependencies

- Updated eclipse-paho-client from 1.2.0 to 1.2.5

# Checklist

- [x] I have updated the relevant documentation to cover the changes implemented. (including changes to the wiki)
- [x] I have specified any additions, alterations or removals of dependencies.
- [x] I have followed the code style present in the project.
- [x] I have updated or added tests as necessary to cover the changes implemented.

### :ballot_box_with_check: Definition of Done checklist
- [x] Meaningful title and valuable description in pull request
- [x] Code change tested manually if not covered by automated tests
- [x] Pull request linked to one or more issues/stories
- [x] Acceptance criteria of the linked stories satisfied
- [x] Code change comprehended by those who approved it
- [x] State of target branch maintained or improved after the code change

### CODE REVIEW

**Comment by Soarnir** [comment]
Do note that this implementation is relatively static, but very easily expandable should we need to cover other android objects.

File: android/Drawer/app/src/main/java/com/example/drawer/MQTTController.java [hunk 1]
```diff
@@ -1,109 +1,165 @@
...
+            });
         } catch (MqttException e) {
             //Standard error printing
-            System.out.println("Could not connectd");
-            System.out.println("reason " + e.getReasonCode());
-            System.out.println("msg " + e.getMessage());
-            System.out.println("loc " + e.getLocalizedMessage());
-            System.out.println("cause " + e.getCause());
-            System.out.println("excep " + e);
+            Log.d(ETAG, "Could not connect");
```
Comment (Kev049) [inline android/Drawer/app/src/main/java/com/example/drawer/MQTTController.java:76]: I understand that logging can be quite beneficial while debugging but the number of Log.d calls should be reduced here.

File: android/Drawer/app/src/main/java/com/example/drawer/MQTTController.java [hunk 2]
```diff
@@ -1,109 +1,165 @@
...
+            Log.d(SUBTAG, "Subscribed to: " + topic);
         } catch (MqttException e) {
             //Standard error printing
-            System.out.println("Subscription could not be performed");
-            System.out.println("reason " + e.getReasonCode());
-            System.out.println("msg " + e.getMessage());
-            System.out.println("loc " + e.getLocalizedMessage());
-            System.out.println("cause " + e.getCause());
-            System.out.println("excep " + e);
+            Log.d(ETAG, "Subscription could not be performed");
```
Comment (Kev049) [inline android/Drawer/app/src/main/java/com/example/drawer/MQTTController.java:100]: Lines 107 - 112 are another example of excessive logging in my opinion. A similar case occurs towards the end of the publish() and disconnect() methods

**Review by Kev049 — CHANGES_REQUESTED** [review]
As marked in-line, the amount of logging seems a bit excessive and I would recommend cleaning it up further before merging to master

File: android/Drawer/app/src/main/java/com/example/drawer/MQTTController.java (hunk 1 above)
Comment (Soarnir) [reply android/Drawer/app/src/main/java/com/example/drawer/MQTTController.java:76]: Fixed in most recent commit.

File: android/Drawer/app/src/main/java/com/example/drawer/MQTTController.java (hunk 2 above)
Comment (Soarnir) [reply android/Drawer/app/src/main/java/com/example/drawer/MQTTController.java:100]: Fixed in most recent commit.

**Review by Kev049 — APPROVED** [review]
Looks good!

**Pull request merged**